### ISSUE TITLE
Support retrieval of content quads for elements with CSS transformations

### ISSUE THREAD

**Issue by whimboo** [w3c/webdriver-bidi#787]
It would be beneficial to support the retrieval of content quads for elements with CSS transformations applied. Currently, this functionality lacks a uniform W3C specification, leading to inconsistent implementations across browsers, including differences in API naming. Both Chrome and Firefox, however, support content quad retrieval.

The content quad of an element should accurately reflect its current position and size, even during CSS transitions, and should be used instead of `getBoundingClientRect()` to get the actual position and size for performing precise actions on such elements.

Integrating this API directly into WebDriver BiDi would enable clients to use a dedicated command that works consistently across all browsers, eliminating the need for browser-specific implementations via `script.evaluate` or `script.callFunction`.

For Chrome, the [`DOM.getContentQuads()`](https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-getContentQuads) command is available in the Chrome DevTools Protocol, while Firefox implements this as `Element.getBoxQuads()`.

**Comment by OrKoN** [comment]
I believe the difference is that quad returns 4 points whereas the rect returns only 2. So rect is always rectangular, while the quad can reflect the actual shape of the element after the transform. I think it can be still important for hit-testing.

**Comment by whimboo** [comment]
Yes, this is actually for https://developer.mozilla.org/en-US/docs/Web/API/DOMQuad and is important (as I already mentioned) when elements are transformed via CSS. 

It looks like that we maybe should discuss it again today when we are all present?

**Comment by gsnedders** [comment]
> while Firefox implements this as Element.getBoxQuads()

Note this is effectively currently unspecified: https://drafts.csswg.org/cssom-view/#dom-geometryutils-getboxquads (though supported in Firefox behind a flag).

See https://github.com/w3c/csswg-drafts/issues/10968.

**Comment by jgraham** [comment]
https://github.com/w3c/csswg-drafts/issues/10537 is an issue for defining `getBoxQuads`, which is a content API implementing this feature. I think this issue is blocked on having some usable definition of what `getBoxQuads` does, and if that definition is also tied to a DOM API I think it would make sense for clients to use that content API via script rather than having a special API in WebDriver that does the same thing.

**Comment by gsnedders** [comment]
See also https://github.com/w3c/webdriver-bidi/issues/769 / https://github.com/microsoft/playwright/issues/32577:

>  Add support for getting content quads to ensure accurate clicking inside transformed elements. This is particularly relevant for tests involving transformed elements, as seen in tests/page/page-click.spec.ts.

**Comment by sadym-chromium** [comment]
>  it would make sense for clients to use that content API via script rather than having a special API in WebDriver that does the same thing.

@jgraham, there are valid testing scenarios requiring such functionality, while DOM API is not fully specified and is not implemented. I find it's useful to have a dedicated BiDi command to provide quads at least until the DOM API is landed. I guess we can simply reference https://drafts.csswg.org/cssom-view/#geometryutils, which will eventually be merged https://github.com/w3c/csswg-drafts/pull/10538.

**Comment by gsnedders** [comment]
I'd much rather we just push on the CSS WG to actually define this rather than working around the CSS WG and define this ourselves. They know their problem space better than we do. If we need this from the CSS WG, we should engage with them to make sure this happens.

**Comment by gsnedders** [comment]
I've Agenda+'d https://github.com/w3c/csswg-drafts/issues/10537, so hopefully it gets on the agenda for [next week](https://www.w3.org/events/meetings/f7fc20fd-5d9e-4427-94cd-34268db72ca8/20250521T120000/) (see there for W3C MO joining instructions).

**Comment by mathiasbynens** [comment]
Has there been a decision here?

**Comment by jgraham** [comment]
AFAICT this hasn't been scheduled at any of the recent CSSWG telecons. I guess it makes sense to talk to some CSSWG members directly to find out what's required to make progress here.

**Comment by whimboo** [comment]
Good news! After a bit of poking once in a while to see updates for the CSSWG PR (https://github.com/w3c/csswg-drafts/pull/10538) I can tell that it just got merged. That means we are no longer blocked and could decide how to make progress on this particular proposed API.

**Comment by whimboo** [comment]
> For Chrome, the [`DOM.getContentQuads()`](https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-getContentQuads) command is available in the Chrome DevTools Protocol, while Firefox implements this as `Element.getBoxQuads()`.

One point we need to discuss is whether we want a dedicated API for this DOM-related command, potentially as the first command in a new `dom` module, or whether clients should be expected to use script evaluation to retrieve the relevant values.

As I understand it, Chrome’s `DOM.getContentQuads` implementation returns final values relative to the top-level browsing context, meaning it automatically accounts for DOM tree traversal. That would not be possible without a dedicated API, and requiring clients to perform all of that computation themselves may not be ideal.

**Comment by OrKoN** [comment]
> As I understand it, Chrome’s DOM.getContentQuads implementation returns final values relative to the top-level browsing context, meaning it automatically accounts for DOM tree traversal.

I think it is relative to the local root frame. I.e., for documents in a cross-process frame, it will be relative to that frame.

**Comment by whimboo** [comment]
Just to highlight the result of our discussion in the last W3C meeting:

> <tidoust> jgraham: It sounds like we should experiment with just the DOM API for now and see whether that's enough to address most of the use cases.

**Comment by hbenl** [comment]
[This commit](https://github.com/hbenl/playwright/commit/47e47bf921483e37a3df13700fc25446d01718fa) shows how the DOM API could be used to retrieve content quads in Playwright. I haven't found any issues with it.

**Comment by jogibear9988** [comment]
getBoxQuads is now in chrome behind a flag (experimental web platform features), and also there is a polyfill for the api (https://github.com/jogibear9988/getBoxQuadsPolyfill)
